apps/src/JSDebuggerUI.js

```diff
@@ -0,0 +1,502 @@
+/** @file Debugger controls and debug console used in our rich JavaScript IDEs */
+/* jshint
+ funcscope: true,
+ newcap: true,
+ nonew: true,
+ shadow: false,
+ unused: true,
+ eqeqeq: true,
+
+ maxlen: 90,
+ maxparams: 3,
+ maxstatements: 200
+ */
+'use strict';
+
+var constants = require('./constants');
+var DebugArea = require('./DebugArea');
+var dom = require('./dom');
+var JSInterpreter = require('./JSInterpreter');
+var Slider = require('./slider');
+var utils = require('./utils');
+
+var _ = utils.getLodash();
+var KeyCodes = constants.KeyCodes;
+var StepType = JSInterpreter.StepType;
+
+/** @const {number} */
+var MIN_DEBUG_AREA_HEIGHT = 120;
+/** @const {number} */
+var MAX_DEBUG_AREA_HEIGHT = 400;
+
+/**
+ * Debugger controls and debug console used in our rich JavaScript IDEs, like
+ * App Lab, Game Lab, etc.
+ * @param {!function} getJSInterpreter - Must be a function that returns the
+ *        current active interpreter, or a falsy value if no interpreter is
+ *        running.
+ * @param {!function} runApp - callback for "launching" the app, which is used
+ *        by the "Step In" button when the app isn't running.
+ * @constructor
+ */
+var JSDebuggerUI = module.exports = function (getJSInterpreter, runApp) {
+
+  /**
+   * Function for getting the active JSInterpreter (which may get replaced on a
+   * regular basis, i.e. for each run).  Should return undefined/null if no
+   * interpreter is currently running.
+   * @private {function}
+   */
+  this.getJSInterpreter_ = getJSInterpreter;
+
+  /**
+   * Callback for "launching" the app, used by the "Step In" button when the app
+   * isn't currently running.
+   * @private {function}
+   */
+  this.runApp_ = runApp;
+
+  /**
+   * Helper that handles open/shut actions for debugger UI
+   * @private {DebugArea}
+   */
+  this.debugOpenShutController_ = null;
+
+  /**
+   * Root element for debug UI: div#debug-area
+   * @private {HTMLDivElement}
+   */
+  this.rootDiv_ = null;
+};
+
+/**
+ * Generate DOM element markup from an ejs file for the debug area.
+ * @param {!function} assetUrl - Helper for getting asset URLs.
+ * @param {!boolean} showButtons - Whether to show the debug buttons
+ * @param {!boolean} showConsole - Whether to show the debug console
+ * @returns {string} of HTML markup to be embedded in page.html.ejs
+ */
+JSDebuggerUI.prototype.getMarkup = function (assetUrl, showButtons, showConsole) {
+  return require('./JSDebuggerUI.html.ejs')({
+    assetUrl: assetUrl,
+    debugButtons: showButtons,
+    debugConsole: showConsole
+  });
+};
+
+/**
+ * Element getter for elements within the debugger UI, which caches its results
+ * so that it's cheap to use over and over.
+ * @type {Function}
+ * @private
+ * @param {string} selector
+ * @returns {HTMLElement}
+ */
+JSDebuggerUI.prototype.getElement_ = _.memoize(function (selector) {
```

Comment:
if you were worried about `initializeAfterDOMCreated` being called twice, you could assign the result of `_.memoize` to a member of `JSDebuggerUI` so that you could reset it on the second call to `initializeAfterDOMCreated`